### ISSUE TITLE
Begin support for Agents

### ISSUE THREAD

**Issue by brainlid** [brainlid/langchain#89]
Agents can act autonomously and perform more complex, multi-step activities.

**Comment by poffdeluxe** [comment]
I've spent some time building a graph-style agent framework with executor on top of this langchain library. Let me know if its something you'd like a demo on and we could possibly discuss how to integrate it

**Comment by brainlid** [comment]
I'm definitely interested in what you've built. 

What do you mean by a graph style agent? I've researched the way the TS LangChain library does agents but I wouldn't describe it as graph related.

**Comment by poffdeluxe** [comment]
So, what I've been working on is more akin to the [LangGraph libraries](https://langchain-ai.github.io/langgraphjs/)

For example, in my work, a simple agent that might do some self-reflection to write an essay is defined like this:
```elixir
  %GraphAgent{
      name: "Essay Writer",
      final_output_property: :latest_revision,
      state: %EssayWriterState{}
    }
    |> GraphAgent.add_node(:first_draft, write_first_draft_node)
    |> GraphAgent.add_node(:write, write_node)
    |> GraphAgent.add_node(:provide_feedback, feedback_node)
    |> GraphAgent.set_entry_point(:first_draft)
    |> GraphAgent.add_edge(:first_draft, :provide_feedback)
    |> GraphAgent.add_edge(:provide_feedback, :write)
    |> GraphAgent.add_conditional_edges(:write, [:end, :provide_feedback], should_continue)
```

Currently, each node is a function that expects a chain in one argument and the input state. Each node returns the new output state

**Comment by nileshtrivedi** [comment]
I have been porting Microsoft's Autogen to Elixir [here](https://github.com/nileshtrivedi/autogen). The goal will be to stay close to Autogen's [conversation patterns](https://microsoft.github.io/autogen/docs/tutorial/conversation-patterns), such as resumable GroupChats and Nested chats.

**Comment by dkuku** [comment]
I think an agent should behave similary to a genserver or liveview:
You would initialize it with some state variables and messages list and then have implemented `handle_*` type callbacks on chain completed event that would receive the response and previous state and set up the state for next chain. Something like:
```elixir
def handle_chain(:chain_x, %{response: response}, state)
   state = assign(state, :chain_x_output, response.content)
  state = assign(state, :messages, state.messages ++ [Message.new_human!(PromptTemplate("the response was <%= chain_x_output %>", state)])
  {:reply, state}
end
```

**Comment by nileshtrivedi** [comment]
I have now started using `langchain` in my [multi-agent library autogen](https://github.com/nileshtrivedi/autogen).

One limitation I ran into was inability to keep track of sender and recipient agents for each message. I am currently using the `name` field for tracking the sender agent, but there's no place for recipient. It would be great if `LangChain.Message` provides a placeholder map for metadata like this. This should be helpful for all agent frameworks that want to use this library.

**Comment by nileshtrivedi** [comment]
@brainlid What's the expected 0.3.0 release timeline? The last RC was back in June.
If the APIs are reasonably stable, perhaps you can cut a second RC release now to build agents on top of?

**Comment by brainlid** [comment]
Yes, I need to cut a new RC release. There's one more breaking change I want to get in before v0.3. At the moment I've very focused on launching a new business/service/marketing so I've gotten behind a bit here.

**Comment by ospaarmann** [comment]
Hey, @brainlid, first of all, thank you for your work. I'm considering using this library in our project, and I would like to understand if this is still actively maintained. Do you still have time, or is your business consuming all of it? :)

**Comment by brainlid** [comment]
Yes, I’m actively working on the library. You can see the work on the main
branch. I needs to cut a new release though.

**Comment by brainlid** [comment]
To be clear @ospaarmann, I am not actively working on this agent feature at the moment, but the library as a whole, yes.

**Comment by gawry** [comment]
> So, what I've been working on is more akin to the [LangGraph libraries](https://langchain-ai.github.io/langgraphjs/)
> 
> For example, in my work, a simple agent that might do some self-reflection to write an essay is defined like this:
> 
>   %GraphAgent{
>       name: "Essay Writer",
>       final_output_property: :latest_revision,
>       state: %EssayWriterState{}
>     }
>     |> GraphAgent.add_node(:first_draft, write_first_draft_node)
>     |> GraphAgent.add_node(:write, write_node)
>     |> GraphAgent.add_node(:provide_feedback, feedback_node)
>     |> GraphAgent.set_entry_point(:first_draft)
>     |> GraphAgent.add_edge(:first_draft, :provide_feedback)
>     |> GraphAgent.add_edge(:provide_feedback, :write)
>     |> GraphAgent.add_conditional_edges(:write, [:end, :provide_feedback], should_continue)
> Currently, each node is a function that expects a chain in one argument and the input state. Each node returns the new output state.
> 
> This style of agent implementation feels really nice and natural in Elixir!

I've kinda started to work on the same thing. How far have you gone with this?

**Comment by poffdeluxe** [comment]
@gawry I have a fairly good first pass working. You can check out the work here: https://github.com/poffdeluxe/magus
I also have an example liveview application that uses the library: https://github.com/poffdeluxe/magus-live

Both should be very functional but also they have a lot of rough edges. I'd love to get more work done on them at some point but there are only so many hours in the day :) Feel free to open PRs or issues! I'd welcome additional assistance or feedback

**Comment by brainlid** [comment]
Implemented. See the project [Sagents](https://github.com/sagents-ai/sagents) project.
